### ISSUE TITLE
Make a plan for sort_values/set_index

### ISSUE THREAD

**Issue by mrocklin** [rapidsai/cudf#2272]
It would be nice to be able to use the `set_index` method to sort the dataframe by a particular column.

There are currently two implementations for this, one in dask.dataframe and one in dask-cudf which uses a batcher sorting net.  While most dask-cudf code has been removed in favor of the dask.dataframe implementations this sorting code has remained, mostly because I don't understand it fully, and don't know if there was a reason for this particular implementation.

Why was this implementation chosen?  Was this discussed somewhere?  Alternatively @sklam, do you have any information here?

cc @kkraus14 @randerzander

**Comment by mrocklin** [comment]
Searchsorted work is happening here: https://github.com/rapidsai/cudf/pull/2156

**Comment by rjzamora** [comment]
Just a minor update here - After [#2156](https://github.com/rapidsai/cudf/pull/2156) goes through, the next issue we run into is the very next line of `set_partitions_pre`:

```
~/workspace/cudf-dask-devel/dask/dask/dataframe/shuffle.py in set_partitions_pre(s, divisions)
    546
    547 def set_partitions_pre(s, divisions):
    548     partitions = pd.Series(divisions).searchsorted(s, side='right') - 1
--> 549     partitions[(s >= divisions[-1]).values] = len(divisions) - 2
    550     return partitions

AttributeError: 'Series' object has no attribute 'values'
```

Since `values` is not a `cudf.Series` property, perhaps the solution is as simple as making it one (and just returning the column values on host)?

**Comment by rjzamora** [comment]
Adding a simple `values` property to `cudf.Series` does seem to get us through the small test.
Changes to `cudf/python/cudf/cudf/dataframe/series.py`:

```python
@property
def values(self):
    return self._values

@property
def _values(self):
    return self.to_pandas().values
```

Test:

```python
In [1]: import cudf, dask.dataframe as dd 
   ...:  
   ...: # Make a dataframe, we'd like to divide the data by y 
   ...: df = cudf.DataFrame({'x': [1, 2, 3, 4, 5, 6], 'y': [0, 4, 0, 4, 0, 4]}) 
   ...:  
   ...: # Split it up into a few partitions with Dask 
   ...: ddf = dd.from_pandas(df, npartitions=3) 
   ...:  
   ...: # Try to create a new dataframe with two partitions sorted on the y column, split by y 0->2 and 2->4 
   ...: out = dd.shuffle.rearrange_by_divisions(ddf, column='y', divisions=[0, 2, 4], shuffle='tasks') 
   ...:  
   ...: # compute in a single thread so it's easy to use %pdb and %debug 
   ...: print(out.compute(scheduler='single-threaded'))                                                                                                                                  
   x  y
0  1  0
2  3  0
4  5  0
1  2  4
3  4  4
5  6  4
```

@kkraus14 - Is there a reason we might want to avoid adding a `values` property to Series (especially one that returns a non-device array)?

**Comment by ayushdg** [comment]
@rjzamora Just tried a small example of `set_index` this morning and saw the same issue of missing the `values` property. #2395

**Comment by jrhemstad** [comment]
> @kkraus14 - Is there a reason we might want to avoid adding a `values` property to Series (especially one that returns a non-device array)?

So copying an entire column from device to host? That would be very expensive.

**Comment by ayushdg** [comment]
>     548     partitions = pd.Series(divisions).searchsorted(s, side='right') - 1
One issue I see above is a hard dependency on pandas. 

>  `549     partitions[(s >= divisions[-1]).values] = len(divisions) - 2`

If partitions were to be a cudf series instead of a pandas series , `values` would not need to be a non-device array

**Comment by rjzamora** [comment]
@jrhemstad  Sorry - That was a silly question!

@ayushdg - Right.  I guess I am just unsure of the best way (if possible) to handle both pandas and cudf using the same logic in Dask, but I haven't exactly tried much yet :)

**Comment by mrocklin** [comment]
Some options to explore:

1.  Can `.values` return something like a cupy array (if it is installed)?
2.  Do we strictly need to use `.values`?  Is this because the result of `searchsorted` is a numpy array rather than a Series?  

    ```python
    def set_partitions_pre(s, divisions):
        partitions = pd.Series(divisions).searchsorted(s, side='right') - 1
        partitions[(s >= divisions[-1]).values] = len(divisions) - 2
        return partitions
    ```

I agree that it would be good not to call `.to_pandas()` here and try to keep things on the device.

See also some conversation about `Series.values` here: https://github.com/rapidsai/cudf/issues/1824

**Comment by ayushdg** [comment]
Also see: #2373 which attempts to add `.values` support by moving to host. Not sure if it's the best option. There is some discussion echoing this point on that pr as well.

**Comment by rjzamora** [comment]
@mrocklin I explored some options to avoid copying the entire column from device to host.  I made [this simple gist](https://gist.github.com/rjzamora/9935311738f70715d7a459e0b1686145) with some experiments.

For a dataframe with 1e8 rows, the dask+pandas version takes about 13.5s, while the dask+cudf version takes about 7.6s.   If I use a `to_pandas()`-based `values` property, the operation takes ~11s

Note that the gist also shows that the cudf version of `searchsorted` is a bit different than the pandas version (it returns a series, rather than a numpy array).  However, this seems reasonable to me..

**Comment by kkraus14** [comment]
7.6s sounds very high, what's taking the time in this situation? Could you share a profile?

**Comment by rjzamora** [comment]
> 7.6s sounds very high, what's taking the time in this situation? Could you share a profile?

@kkraus14 agreed - I'll take a closer look and collect a profile

**Comment by mrocklin** [comment]
> Note that the gist also shows that the cudf version of searchsorted is a bit different than the pandas version (it returns a series, rather than a numpy array). However, this seems reasonable to me..

So lets take a look at the `set_partitions_pre` function for a moment

https://github.com/dask/dask/blob/e0a77233fd4c5c3fc0ea8d533145778289c69078/dask/dataframe/shuffle.py#L547-L550

```python
def set_partitions_pre(s, divisions):
    partitions = pd.Series(divisions).searchsorted(s, side="right") - 1
    partitions[(s >= divisions[-1]).values] = len(divisions) - 2
    return partitions
```

The reason that `.values` has come up is in the second line, where presumably we're trying to index into `partitions` with something.  If it's a numpy array (as is the case when `s` is a pandas.Series) then we apparently need a `.values` call (although I would be a surprised if NumPy couldn't figure things out here).  (for background, I think that this is saying "Hey! Every row that is targetting something beyond the last parittion, go to the last partition instead!"  I have no idea why we need that, but there is probably some test for it that would fail that comes up in some corner case (this is worth checking to make sure though).  

So, as you say, we might have a type issue in that cudf returns a series rather than an array-like.  This seems to me, as you say, a sensible choice of cudf's part.  So if we focus on just this function we might do a few things:

1.  Implement `.values` and coerce things to an array-like
2.  Not implement `.values` but keep things as a series all the way through, perhaps with an `if is_arraylike(partitions)` branch
3.   ??

To figure out the right approach we probably need to zoom out a bit and look at the context in which this is called, and what happens next to these objects.  Are the next operations also array/dataframe agnostic, or will we need to coerce at some point anyway?

It looks like the next thing that happens is that they get assigned back into the dataframe.  So if we go the array route it would be good to ensure that this is possible.  If we go the series route then we might not care.  

https://github.com/dask/dask/blob/e0a77233fd4c5c3fc0ea8d533145778289c69078/dask/dataframe/shuffle.py#L180

```python
        df2 = df.assign(_partitions=partitions)
```

I'm curious how much works if we add an `is_arraylike` condition like the following:

```python
def set_partitions_pre(s, divisions):
    partitions = pd.Series(divisions).searchsorted(s, side="right") - 1
    index = s > divisions[-1]
    if is_arraylike(partitions):
        index = index.values
    partitions[index] = len(divisions) - 2
    return partitions
```

Or if we can remove the `.values` call entirely, or if this causes something to break, what is it, and should we fix that instead?

```python
def set_partitions_pre(s, divisions):
    partitions = pd.Series(divisions).searchsorted(s, side="right") - 1
    partitions[s > divisions[-1]] = len(divisions) - 2
    return partitions
```

**Comment by rjzamora** [comment]
Thanks for the feedback @mrocklin!  Just to provide a bit more information: For the "cupy" experiment I shared, I am using a custom `set_partitions_pre_cupy_series` function, which takes in divisions as a series (e.g. cupy.Series):

```
def set_partitions_pre_cupy_series(s, divisions):
    partitions = (divisions.searchsorted(s, side="right") - 1).values_cupy
    partitions[(s >= divisions.iloc[-1]).values_cupy] = len(divisions) - 2
    return partitions
```

This ultimately returns partitions as a cupy array.  To do this, I added a (rough) `values_cupy` property to Series:

```
@property
def values_cupy(self):
     try:
         import cupy
         return cupy.asarray(self._column._data.mem)
    except ImportError:
        return self.to_array()
```

Overall, I am thinking it is likely best to do something like the `is_arraylike(partitions)` branch you mentioned.  It may also make sense to remove the explicit pandas dependency by changing the divisions input type to be a series.   The question I am struggling with at the moment is how to efficienly reset the `partitions` elements in a general way

> I have no idea why we need that, but there is probably some test for it that would fail that comes up in some corner case (this is worth checking to make sure though).

This actually happens for every value belonging to the last partition here.  So, might not be much of a corner case.

Regarding the `values` property: [#2373](https://github.com/rapidsai/cudf/pull/2373) added this today.  So, we don't have to worry about implementing a host-based `values` definition if it is needed.

Regarding performance:

As far as I can tell, a good chunk of time is likely python/dask.  Otherwise, there seems to be a good chunk of time (~45%) in Dask's `shuffle_group`, which does make some pandas and numpy-specific calls.

**Comment by rjzamora** [comment]
Just a followup - The test I described above uses cupy to calculate the new partition of each row.  Unfortunately, the next phase of actually sorting the dataframe into these new partitions requires everything to be copied to host anyway.

The actual splitting of the dataframe into new partitions (within each of the original partitions) is accmplished in `shuffle_group` (and `shuffle_group_2`).  My profiling indeed shows that these functions account for much of the `rearrange_by_divisions` call.  Here is `shuffle_group` (`shuffle_group_2` is similar):

```python
def shuffle_group(df, col, stage, k, npartitions):
    """ Splits dataframe into groups ..."""
    if col == "_partitions":
        ind = df[col] # True in our case
    else:
        ind = hash_pandas_object(df[col], index=False)

    c = ind._values  # Need `_values` property in cudf series for this
    typ = np.min_scalar_type(npartitions * 2)

    c = np.mod(c, npartitions).astype(typ, copy=False)
    np.floor_divide(c, k ** stage, out=c)
    np.mod(c, k, out=c)

    indexer, locations = groupsort_indexer(c.astype(np.int64), k)  # Numpy/pandas specific
    df2 = df.take(indexer)
    locations = locations.cumsum()
    parts = [df2.iloc[a:b] for a, b in zip(locations[:-1], locations[1:])]

    return dict(zip(range(k), parts))
```

Like `set_partitions_pre`, this function introduces a couple issues for cudf:

- We are assuming a `_values` property exists for the series/column being used for partitioning.  This is not true for cudf (I believe even #2373 left this out).
- Pandas' `groupsort_indexer` function requires that a numpy array be input for the first argument.  This is fine if we return a numpy array for `_values`, but not if we use a zero-copy cupy array

Overall, [the `groupsort_indexer` function](https://github.com/pandas-dev/pandas/blob/c7a1321029e07ee6d7ea30036649b488b2e362f7/pandas/_libs/algos.pyx#L151) is pretty simple.  It takes in an array of groups (partition numbers in our case), and the total number of groups (number of partitions).  The output is the new group-sorted index, and an array with the number of items in each group.   @kkraus Does a similar algorithm already exist in cudf?

**Comment by mrocklin** [comment]
cc'ing @kkraus14 (note the 14 in his username)

Repeating the question here:

> Overall, the groupsort_indexer function is pretty simple. It takes in an array of groups (partition numbers in our case), and the total number of groups (number of partitions). The output is the new group-sorted index, and an array with the number of items in each group. @kkraus Does a similar algorithm already exist in cudf?

**Comment by mrocklin** [comment]
I think that in private conversation Keith also mentioned that there may be
some operation in cudf that does several of these steps all at once, which
may be another approach

**Comment by kkraus14** [comment]
We currently only have a hash partitioner function which does the equivalent of hash the input dataframe based on some key column(s), run a modulo against the number of partitions N, split the dataframe into N dataframes based on the hash % N.

This doesn't seem quite the same, no?

**Comment by mrocklin** [comment]
No, not quite the same.  

At this point we have a dataframe with a new column that has the output partition of each row.  We now want to split our dataframe up into many small dataframes, one for each output partition.

We do this by creating an indexer which will effectively sort things by that column, and then calling `take` to do the reordering, and then calling `iloc` many times to split things apart (though this could be a bulk operation in preferred

```python
    c = ... # the output partition locations
    indexer, locations = groupsort_indexer(c.astype(np.int64), k)  # link to this above in Rick's comment
    df2 = df.take(indexer)
    locations = locations.cumsum()
    parts = [df2.iloc[a:b] for a, b in zip(locations[:-1], locations[1:])]
```

**Comment by jrhemstad** [comment]
> At this point we have a dataframe with a new column that has the output partition of each row. We now want to split our dataframe up into many small dataframes, one for each output partition.

So in `libcudf` we already have a `gdf_hash_partition` algorithm that does this, but uses the hash values of DF rows to determine the partition mapping.

It would appear that here you're instead using an arbitrary column of values to determine the partition mapping. 

It would be fairly trivial to split up and generalize the `gdf_hash_partiton` function to do what you need here. 

All of this code could be replaced with a single libcudf function that I'm guessing will be orders of magnitude faster. 

```
    c = np.mod(c, npartitions).astype(typ, copy=False)
    np.floor_divide(c, k ** stage, out=c)
    np.mod(c, k, out=c)

    indexer, locations = groupsort_indexer(c.astype(np.int64), k)  # Numpy/pandas specific
    df2 = df.take(indexer)
    locations = locations.cumsum()
    parts = [df2.iloc[a:b] for a, b in zip(locations[:-1], locations[1:])]
```

**Comment by mrocklin** [comment]
Right, we're being a little bit careful about where things go.  Sometimes
we move them around in multiple stages, or have particular destinations in
mind based on values.

Putting all of those lines into a single operation sounds fine from my
perspective.

**Comment by jrhemstad** [comment]
> Right, we're being a little bit careful about where things go. Sometimes we move them around in multiple stages, or have particular destinations in mind based on values. Putting all of those lines into a single operation sounds fine from my perspective.

I'm a bit confused because that sounds different than what you originally described. 

To be clear, what I'm envisioning is an algorithm that does the following:

Inputs:
- A DataFrame with `N` rows
- A `partition_map` of `N` values from `[0, m)`

Outputs:
- `m-1` DataFrames where DF `i` contains all of the rows from the input DF where `partition_map[j] == i`

**Comment by mrocklin** [comment]
That sounds good to me

**Comment by jrhemstad** [comment]
Out of curiosity, how is the `partition_map` being generated for this particular use case? Obviously it has to do with which partition a row maps to based on the sorting, but I'm curious how that's being determined.

**Comment by rjzamora** [comment]
Thanks for the input @jrhemstad @kkraus14 @mrocklin!  This is super useful.   Sorry - I didn't see this recent discussion until now.

**Comment by rjzamora** [comment]
> Out of curiosity, how is the partition_map being generated for this particular use case? Obviously it has to do with which partition a row maps to based on the sorting, but I'm curious how that's being determined.

If I follow the question correctly, the partition map (`partitions`) is currently determined using `searchsorted`:

```python
partitions = pd.Series(divisions).searchsorted(s, side="right") - 1
```
where `divisions` is a list of divisions along the index/column (`s`) that we want to partition on.

**Comment by jrhemstad** [comment]
So it sounds like the size of `partitions` is equal to the number of desired partitions. That's not quite the same thing as the `partition_map` I described, which is equal in size to the number of rows in the DF being partitioned. 

How do you determine the mapping of each row in the DF to a particular partition? (this is what the `partition_map` I was describing is)

**Comment by rjzamora** [comment]
In the case I described, the size of `partitions` is indeed equal to the number of rows in the DF being partitioned, because `s` corresponds to a full column.

**Comment by jrhemstad** [comment]
What is the size of `divisions`? I think I'm misunderstanding how `searchsorted` works.

**Comment by rjzamora** [comment]
Thats right - `divisions` is just a list with length (`npartitions` + 1).  There are ways a partition can be calculated, but you can really just think of partitions as a user input for now.  For example, the user can specify that they want to partition along some column `y` using `divisions = [0,0.3,0.6]`.  

To do this, they might call:

```python
df  = pd.DataFrame( {'x': [i for i in range(size)], 'y': np.random.randn(size)} )
ddf = dd.from_pandas(df, npartitions=3)

out = dd.shuffle.rearrange_by_divisions(ddf, column='y', divisions=[0,0.3,0.6], shuffle='tasks')
result_df = out.compute()
```
`rearrange_by_divisions` will start by calculating the `partition_map` (`partitions`) you described (using `search_sorted`). The function looks like this:

```python
def set_partitions_pre(s, divisions):
    partitions = pd.Series(divisions).searchsorted(s, side="right") - 1
    partitions[(s >= divisions[-1]).values] = len(divisions) - 2
    return partitions
```

After we have `partitions`, we need call `rearrange_by_column` to perform the actual shuffling (either in disk or in memory). On a high level, the in-memory version (rearrange_by_column_tasks) does two things:

- Uses `shuffle_group` within each of the old partitions to split/sort the local dataframe according to the new partitions
- Concatenates the resulting splits into the new divisions

My understanding is that the number of partitions is typically orders of magnitude smaller than the size of the dataframe in practice (in order for each worker/task to have sufficient data/computation for efficiency)

@mrocklin feel free to correct me or expand.

**Comment by jrhemstad** [comment]
Thanks @rjzamora that's very helpful.

I'm trying to think what the right primitive(s) is for libcudf/cuDF to expose to accelerate this operation. 

We have a few options:
1. We could make a primitive that could entirely replace `rearrange_by_divisions`
   - You give us a DF and a set of divisions, and we'll partition up the DF 
   - This would likely be the fastest
2. We could keep Dask's `rearrange_by_division`, and instead replace `shuffle_group`
   - That is effectively what I proposed here: https://github.com/rapidsai/cudf/issues/2452
   - This is similar to what we did with `gdf_hash_partition` for `dask-cudf`'s distributed Join and it was very effective, see https://github.com/dask/dask/issues/4876
   - Dask would still compute the `partition_map`, but cuDF would handle doing the actual partitioning
   - This would be the more general option 

Curious what @kkraus14 @harrism think the right primitive to expose is here.

**Comment by rjzamora** [comment]
I might be missunderstanding,  but since `rearrange_by_column_tasks` operates on a *dask* dataframe, the libcudf/cuDF primitives would most likely need to be more fine-grained (along the lines of option 2).  It would be great to be wrong about this :)

From my perspective, it seems useful to replace the `set_partitions_pre` function (shown above) with a libcudf/cuDF primitive, as well as `shuffle_group`, and possibly `shuffle_group_2`.  The latter two are both based on pandas' `groupsort_indexer` at the moment.  I will look through these functions in detail to get a better idea of if/how the logic can be combined.

**Comment by mrocklin** [comment]
In order to avoid centralized overhead, dask.dataframe will sometimes move things around in a couple of stages.  What we need is the in-memory operation that occurs in each stage.  This operation is `shuffle_group`.  Note that it includes a bit of finicky math here:

```python
    c = ind._values
    typ = np.min_scalar_type(npartitions * 2)

    c = np.mod(c, npartitions).astype(typ, copy=False)
    np.floor_divide(c, k ** stage, out=c)
    np.mod(c, k, out=c)
```

And then the actual splitting

```python
    indexer, locations = groupsort_indexer(c.astype(np.int64), k)
    df2 = df.take(indexer)
    locations = locations.cumsum()
    parts = [df2.iloc[a:b] for a, b in zip(locations[:-1], locations[1:])]

    return dict(zip(range(k), parts))
```

I think that a single function in cudf that did all of this would be great.  I would also understand if cudf only wanted to implement the second part, which is a bit more generally useful.

**Comment by jrhemstad** [comment]
@mrocklin can you explain the finicky math part? 

It sounds like the partition mapping returned by `set_partitions_pre` isn't the final mapping and requires some additional modification.

**Comment by mrocklin** [comment]
I've explained the motivation here as a PR with a nicer docstring.  If you
need the specific approach let me know.  It'll probably take some time to
write down, so I haven't done it yet.  If you're just curious here then
I'll pass for now in the interests of time.  If it's important for solving
this problem then I'll prioritize this appropriately.

Next time we're together perhaps we can find a whiteboard, go through it,
and then I can pester you to write it up into docs somewhere?  :)

**Comment by mrocklin** [comment]
I forgot to link to the PR: https://github.com/dask/dask/pull/5213

**Comment by jrhemstad** [comment]
> I think that a single function in cudf that did all of this would be great.

I can't develop a single function to do `shuffle_group` if I don't understand the finicky math part. 

So the best I can do is what I described here: https://github.com/rapidsai/cudf/issues/2452 which requires you to materialize the final `partition_map` for a DF.

**Comment by mrocklin** [comment]
Yeah, that looks great.  That's effectively the second part of the two parts described here: https://github.com/rapidsai/cudf/issues/2272#issuecomment-517727092

```python
    indexer, locations = groupsort_indexer(c.astype(np.int64), k)
    df2 = df.take(indexer)
    locations = locations.cumsum()
    parts = [df2.iloc[a:b] for a, b in zip(locations[:-1], locations[1:])]

    return dict(zip(range(k), parts))
```

The first part, the finicky path part ... 

```python
    c = ind._values
    typ = np.min_scalar_type(npartitions * 2)

    c = np.mod(c, npartitions).astype(typ, copy=False)
    np.floor_divide(c, k ** stage, out=c)
    np.mod(c, k, out=c)
```

We can probably just do with cupy, assuming that we can get a device-array-like out of a cudf series, perhaps with a `.values` attribute that does this.

If you wanted to combine the two steps, that would be fine with me, my sense is that we should keep them separate first for the sake of generality, and only proceed to merge them if it's a large bottleneck (or if it's hard for cudf to implement a `.values` attribute that generates a device-array-like object (which may be the case))

I don't think that you need to actually understand the finicky math here in order to implement it.  The `c` attribute is something like your `partition_map`.  We can't go directly from input partition to output partition all in one go because that generates millions of tasks.  So we do this in stages.

**Comment by rjzamora** [comment]
> We can probably just do with cupy, assuming that we can get a device-array-like out of a cudf series, perhaps with a .values attribute that does this.

Just a note that cupy can (as far as I can tell) be used to produce a device array for the values of a cudf series.  I've already tested this by tweaking `set_partitions_pre` to do something like this:

```python
def set_partitions_pre_cupy(s, divisions):
    partitions = (cudf.Series(divisions).searchsorted(s, side="right") - 1).values_cupy
    partitions[(s >= divisions[-1]).values_cupy] = len(divisions) - 2
    return partitions
```
where the `values_cupy` property is just returning `cupy.asarray(self._column._data.mem)`

**Comment by kkraus14** [comment]
> We can probably just do with cupy, assuming that we can get a device-array-like out of a cudf series, perhaps with a `.values` attribute that does this.

If we use CuPy here how do we handle nulls? We can't.

**Comment by mrocklin** [comment]
You have `.values` raise NotImplementedError if there are nulls.

In this case, we won't have any nulls.  In the future, maybe we'll figure
out something better.

**Comment by kkraus14** [comment]
If we hit the first branch of the shuffle:
```
    if col == "_partitions":
        ind = df[col] # True in our case
```
`ind` can absolutely have nulls here and if it does the codepath with CuPy will break. If we just had `.value` return a cuDF index or a cuDF series we could then use NEP18 to handle the `np.mod` and `np.floordivide` calls following it with null support nicely until we get into the actual data movement for partitioning, where our partitioning functions will handle nulls.

**Comment by mrocklin** [comment]
We've assigned `_partitions` ourselves.  It's a unsigned integer column
with no nulls.

**Comment by kkraus14** [comment]
Okay, so it sounds like if we can resolve the CuPy packaging issues with regards to cuML / NCCL then we can use it for the `.values` return.

Should be relatively straightforward especially in using #2433 to construct the CuPy array with zero copy.

**Comment by mrocklin** [comment]
> Should be relatively straightforward

@kkraus14 no pressure (well, a little) but do you have an expected timeline on this?  This affects some planning on how to handle this issue.

**Comment by rjzamora** [comment]
Just to summarize the status here:

- There is currently  [a PR in Dask](https://github.com/dask/dask/pull/5205) with the goal of supporting cudf-based dask data frames for `rearrange_by_divisions` and `set_index`.  This PR does not aim for an optimal solution, but for simple changes to dask.  Hopefully we can build upon the existing changes there to produce a *real* solution

- Both `rearrange_by_divisions` and `set_index` rely on two general steps: (step 1) The generation of a partition map, which labels the new partition for each row in the dataframe, and (step 2) The actual sorting/rearranging of the data frame into the new partitions

    - The generation of the partition map (step 1 above) relies on the `set_partitions_pre` function

    - The sorting/rearranging (step 2 above) relies on `rearrange_by_column_tasks`, which is often dominated by `shuffle_group` and `shuffle_group_2`. The functions, in turn, use pandas’ `groupsort_indexer` to split each *old* partition into the pieces that belong to each *new* partition

- **The current plan**:  Accelerate the `shuffle_group` functions by providing a gpu-accelerated replacement for `groupsort_indexer`, and to avoid moving data to disk by leaning on cupy when possible...    Do I have this right?

**Comment by mrocklin** [comment]
> Accelerate the shuffle_group functions by providing a gpu-accelerated replacement for groupsort_indexer, and to avoid moving data to disk by leaning on cupy when possible... Do I have this right?

OK, so it looks like we're currently blocked on 

-  a cupy `.values` attribute (probably shortly after https://github.com/rapidsai/cudf/pull/2464)
-  partitioning a table based on a partition-map https://github.com/rapidsai/cudf/issues/2452

@kkraus14 @harrism for scheduling do you have an estimate about when those might finish?  If they're likely to take a while (everyone has lots of work to do) are they things that someone like @rjzamora could solve, or does it make more sense to have someone more familiar with that code handle them?

**Comment by rjzamora** [comment]
Some more information that is likely important, and may motivate a more-complete gpu version of `shuffle_group`...  Here is a snakeviz profile of `shuffle_group` during a `rearrange_by_divisions` operation on a dataframe with 1e7 rows and 3 partitions:

![shuffle_group](https://user-images.githubusercontent.com/20461013/62582051-abd19d80-b870-11e9-9371-5ca32c5643fb.png)

**Notes** (counting rows from top to bottom):
- The dark blue block in the second row is `groupsort_indexer`
- The dark orange block in second row (and below) is cudf indexing (`_getitem_tuple_arg`) within list comprehension in dask

**Comment by kkraus14** [comment]
@rjzamora any chance you can dump an html file that will allow us to explore this data ourselves? Curious to see what else is in this callstack.

**Comment by mrocklin** [comment]
Rather than dumping an HTML file I recommend producing a pstats output file
by using %%prun in a notebook or the cProfile module from the command line
with the -D flag .  That file can then be shared and others can run
snakeviz on that file.

**Comment by mrocklin** [comment]
See
https://ipython.readthedocs.io/en/stable/interactive/magics.html#magic-prun and
-D

**Comment by rjzamora** [comment]
Thanks for following up on this @kkraus14, and thanks @mrocklin for the suggestion - I am about to run some new tests pretty soon.  I will share the raw profile that I collect.

**Comment by rjzamora** [comment]
Okay - I posted a newer raw profile [here](https://github.com/rjzamora/notebooks/blob/master/profiles/rearrange_by_divisions_dask-e4a7c9d.prof) (collected with Python 3.7).

I also have an html version [here](https://github.com/rjzamora/notebooks/blob/master/profiles/rearrange_by_divisions_dask-e4a7c9d.htm) in case the raw profile is problematic.

**Comment by mrocklin** [comment]
We're probably ready to remove some of the older dask-cudf implementations of set_index/merge/join.  I've raised this as a separate issue here: https://github.com/rapidsai/cudf/issues/2598

**Comment by mrocklin** [comment]
I thought I'd poke this.  @harrism @jrhemstad do either of you have thoughts on what a timeline for the partition_map splitting operation could look like?  We're quickly reaching a state where this will be our only remaining blocker.

cc @randerzander

**Comment by harrism** [comment]
@mrocklin yes, @jrhemstad and I have been a plan. A "fastest time to solution, not necessarily fastest solution" plan is described in #2677. Please review.

**Comment by randerzander** [comment]
Now that all of the issues mentioned above have closed, and cupy packaging has been resolved, are we able to continue deprecating the custom dask_cudf sort code?

My hope is doing so will yield performance and scalability improvements. Several workflows have recently had problems with the dask_cudf implementation of set_index and sort_values.

If it's useful, I'm happy to open new issues with example unexpected behavior, but a quick read of the above makes me feel like you all have a good handle on the problem.

**Comment by rjzamora** [comment]
> ...are we able to continue deprecating the custom dask_cudf sort code?... My hope is doing so will yield performance and scalability improvements. Several workflows have recently had problems with the dask_cudf implementation of set_index and sort_values.

@randerzander  Can you clarify the pieces that are still custom to dask_cudf?  dask_cudf should now be using the main-line dask implementation of `set_index`.  Since `sort_values` is not supported in dask, we never really made a plan to remove the cudf-specific implementation.

If your comment is more about performance improvements/optimizations, then I can say we are certainly in the process of exploring ways to (1) Improve merge/join performance to better match [the expected/available performance](https://github.com/rapidsai/ucx-py/pull/378), and (2) improve the stability of operations like `set_index`.

Note that I am currently working on a simple (but experimental) modification to the upstream-dask `shuffle` and `set_partition` code.  It seems that we can use hashing to moderately reduce some of the memory consumption and  and data-copy overhead.

**Comment by harrism** [comment]
To what extent is this issue satisfied by the current work on providing multi-column distributed sorting primitives (table quantiles, k-way merge) in libcudf?

**Comment by kkraus14** [comment]
> To what extent is this issue satisfied by the current work on providing multi-column distributed sorting primitives (table quantiles, k-way merge) in libcudf?

100%, this is the primary target for that work.

**Comment by harrism** [comment]
@kkraus14 both of those new primitives were added. Does this need to stay open?

**Comment by kkraus14** [comment]
I defer to @rjzamora if he'd like to keep this open for future discussion or if this can now be closed.

**Comment by rjzamora** [comment]
We should probably close this once [#4308](https://github.com/rapidsai/cudf/pull/4308) is merged.  That should provide dask_cudf with the general sorting/set_index functionality targeted by this discussion.  After that, the remaining work will be to push as mach dask_cudf code as possible into main-line dask (which can probably be tracked separately).

**Comment by rjzamora** [comment]
Now that #4308  is merged, it probably makes sense to close this.  We can open new issues in the future to discuss any performance/functionality challenges that come up.

Also, the task of moving the new `sort_values` and `repartition_by_hash` functionality into upstream dask can/should be tracked elsewhere.

**Comment by rjzamora** [comment]
@randerzander - Feel free to reopen this if the current `sort_values` solution does not end up meeting your needs.  You can also open a more-targeted issue.
